File: leetcode/algorithms/c#/Linked List Cycle II/Linked List Cycle II.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution
{
    public ListNode DetectCycle(ListNode head)
    {
        Dictionary<ListNode, int> source = new Dictionary<ListNode, int>();
        while (head != null)
        {
            if (source.ContainsKey(head)) return head;
            source.Add(head, 0);
            head = head.next;
        }
        return null;

    }
}

File: leetcode/algorithms/c++/Linked List Cycle II/Linked List Cycle II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        unordered_set<ListNode *> base;
        while (head != NULL)
        {
            if (base.find(head) != base.end())
                return head;
            base.insert(head);
            head = head->next;
        }
        return NULL;
    }
};